// @front.build/backend/plugin/Lambda-build-function/index.js

var fs = require('fs')

/**
 * Build the lambda function on the lambda machine.
 */

module.exports = function(state, action) {
  state.dispatch({
    type: 'Run process',
    input: {
      Actions: [
        {
          type: 'Write file',
          input: {
            Files: [
              {
                'File path': 'index.js',
                'File content': action.input['index.js']
              },
              {
                'File path': 'package.json',
                'File content': action.input['package.json']
              }
            ]
          }
        },
        {
          type: 'NPM install'
        },
        {
          type: 'Run command',
          input: {
            Input: 'chmod 777 node_modules'
          }
        },
        {
          type: 'Run command',
          input: {
            Input: 'zip -ur9 ./plugin.zip index.js package.json node_modules'
          }
        },
        {
          type: 'Run command',
          input: {
            Input: 'chmod 777 plugin.zip'
          }
        },
        {
          type: 'Remove file',
          input: {
            Files: [
              {
                'File path': 'index.js'
              },
              {
                'File path': 'package.json'
              }
            ]
          }
        }
      ]
    }
  })
}
