// @front.build/backend/plugin/Heroku-create-build/index.js

// https://devcenter.heroku.com/articles/setting-up-apps-using-the-heroku-platform-api
// https://devcenter.heroku.com/articles/setting-up-apps-using-the-heroku-platform-api#polling-for-setup-status
// https://devcenter.heroku.com/articles/build-and-release-using-the-api#creating-builds

module.exports = function(state, action, callback) {
  var appId = action.input['App ID']
  var url = action.input['App upload URL']
  var version = action.input['App version']

  // source_blob:url
  // source_blob:checksum
  state.dispatch({
    type: 'Request JSON',
    input: {
      URL: 'https://api.heroku.com/apps/' + appId + '/builds',
      Method: 'POST',
      Body: {
        source_blob: {
          url: url,
          version: version
        }
      },
      Headers: {
        'Accept': 'application/vnd.heroku+json; version=3'
      }
    }
  }, callback)

//   {
//   "app": {
//     "id": "01234567-89ab-cdef-0123-456789abcdef"
//   },
//   "buildpacks": [
//     {
//       "url": "https://github.com/heroku/heroku-buildpack-ruby"
//     }
//   ],
//   "created_at": "2012-01-01T12:00:00Z",
//   "id": "01234567-89ab-cdef-0123-456789abcdef",
//   "output_stream_url": "https://build-output.heroku.com/streams/01234567-89ab-cdef-0123-456789abcdef",
//   "source_blob": {
//     "checksum": "SHA256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
//     "url": "https://example.com/source.tgz?token=xyz",
//     "version": "v1.3.0"
//   },
//   "release": {
//     "id": "01234567-89ab-cdef-0123-456789abcdef"
//   },
//   "slug": {
//     "id": "01234567-89ab-cdef-0123-456789abcdef"
//   },
//   "stack": "heroku-16",
//   "status": "succeeded",
//   "updated_at": "2012-01-01T12:00:00Z",
//   "user": {
//     "id": "01234567-89ab-cdef-0123-456789abcdef",
//     "email": "username@example.com"
//   }
// }
}
